### ISSUE TITLE
[zscale_zia] Validate response format versions

### ISSUE THREAD

**Issue by chrisberkhout** [elastic/integrations#10743]
The `zscale_zia` integration requires the user to set up Zscaler Internet Access (ZIA) with response format definitions for each type of data requested by the integration.

If a user upgrades the integration without updating their response format configuration in Zscaler, there can be a mismatch between the data in responses and what's expected by the ingest pipelines.

We could avoid such problems by having each format definition include a hard-coded version number, and having the ingest pipelines validate that they have received data with the expected version number.

When introducing this, it would be good to allow responses with the correct version number, or with no version number but the expected a set of field names (which could be validated by sorting, hashing and comparing to an expected value). That would avoid having a breaking change to introduce the version validation.

**Comment by chrisberkhout** [comment]
Moved to Quality category from SDH (but both apply).

**Comment by efd6** [comment]
@chrisberkhout Is this the kind of thing you were thinking? I went with a direct comparison rather than a hash since we can use the value in this case as an informative error value, whereas with a hash we cannot.
```diff
diff --git a/packages/zscaler_zia/data_stream/audit/elasticsearch/ingest_pipeline/default.yml b/packages/zscaler_zia/data_stream/audit/elasticsearch/ingest_pipeline/default.yml
index 168217e738..dbdf854e34 100644
--- a/packages/zscaler_zia/data_stream/audit/elasticsearch/ingest_pipeline/default.yml
+++ b/packages/zscaler_zia/data_stream/audit/elasticsearch/ingest_pipeline/default.yml
@@ -37,6 +37,30 @@ processors:
       tag: rename_resp_event
       target_field: json
       ignore_missing: true
+  - script:
+      params:
+        expect:
+          fields: action|adminid|auditlogtype|category|clientip|errorcode|interface|postaction|preaction|recordid|resource|result|subcategory|time|timezone
+          version: 1
+      if: ctx.json != null
+      source: |-
+        if (ctx.json.version == null) {
+          def fields = [];
+          for (e in ctx.json.entrySet()) {
+            fields.add(e.getKey());
+          }
+          Collections.sort(fields);
+          String signature = String.join("|", fields);
+          if (signature != params.expect.fields) {
+            ctx.error = ctx.error ?: [:];
+            ctx.error.message = ctx.error.message ?: [];
+            ctx.error.message.add("field set mismatch: "+signature+" is not expected set of templated fields");
+          }
+        } else if (ctx.json.version != params.expect.version) {
+            ctx.error = ctx.error ?: [:];
+            ctx.error.message = ctx.error.message ?: [];
+            ctx.error.message.add("template version mismatch: "+ctx.json.version.toString()+" is not expected version");
+        }
   - remove:
       field: resp
       tag: remove_resp
```

**Comment by chrisberkhout** [comment]
@efd6 Yes, I was thinking of something like that.

On second thought though, maybe now we can just do the version check.

The current version does involve a field change from previous version, so I think we no longer have the case of an unversioned pattern with all other fields up to date.

It's also less maintenance overhead.

**Comment by efd6** [comment]
> The current version does involve a field change from previous versions, so I think we no longer have the case of an unversioned pattern with all other fields up to date.

Yes, this is essentially what the text of the box above was going to be before I did the work to look into what we should do.

What I _was_ thinking is that we could have two part test as exists in the code above for some Reasonable™ period to pick up straglers, and then remove the fields-based check and rely solely on the version after that period has elapsed.

> It's also less maintenance overhead.

This would be addressed by the transitional approach I have above.

**Comment by chrisberkhout** [comment]
Note: The version number has been present since v8, 2024-08-23.

> What I was thinking is that we could have two part test as exists in the code above for some Reasonable™ period to pick up straglers, and then remove the fields-based check and rely solely on the version after that period has elapsed.

Okay, that's good.

**Comment by efd6** [comment]
> The version number has been present since v8, 2024-08-23.

Do you have a link for that?

**Comment by chrisberkhout** [comment]
> > The version number has been present since v8, 2024-08-23.
>
> Do you have a link for that?

You can see the change in the README here:

https://github.com/elastic/integrations/pull/10855/files#diff-8034be6ad1a0f6185169e497dc6bd22985ac5e28676552f7441d0b2732b8921a

The version number was noted in the README before, but that PR is when it was added inside the pattern.
